sophia/src/triple/stream/_error.rs

```diff
@@ -0,0 +1,214 @@
+use std::convert::Infallible;
+use std::error::Error;
+use std::fmt;
+
+/// A error that is raised by functions that move fallible `Source`s into
+/// fallible `Sinks`.
+///
+/// In case this error is raised it can be matched to investigate if `Source`
+/// or `Sink` failed.
+///
+/// The `std::error::Error::source()` method is implemented accordingly.
+///
+/// # ToDo
+///
+/// For a proper use in multi-threaded environments the trait bounds should be
+/// supplied with `+ Send + Sync`. However, such bounds wound make `StreamError`
+/// incompatible with `sophia`'s whole error-handling. This should be resolved
+/// when the error-handling is completly refactored
+/// ([tracking issue](https://github.com/pchampin/sophia_rs/issues/8)).
+#[derive(Debug)]
+pub enum StreamError<SourceErr, SinkErr>
+where
+    SourceErr: 'static + Error,
+    SinkErr: 'static + Error,
+{
+    Source(SourceErr),
+    Sink(SinkErr),
+}
+
+use self::StreamError::*;
+
+impl<SourceErr, SinkErr> StreamError<SourceErr, SinkErr>
+where
+    SourceErr: 'static + Error,
+    SinkErr: 'static + Error,
+{
+	/// Build a `StreamError` from an error of the `Source`.
+    pub fn from_source_err(se: SourceErr) -> Self {
+        Source(se)
+	}
+	/// Build a `StreamError` from an error of the `Sink`.
+    pub fn from_sink_err(se: SinkErr) -> Self {
+        Sink(se)
+    }
```

Comment:
I'm really not sure we need this if we publicly expose SourceError and SinkError...
`SourceError(err)` is much easier than `StreamError::from_source_err(err)`